### PR TITLE
ORC-346: [C++] Add one second when writing negative Timestamp's with non zero nanos …

### PR DESCRIPTION
**Pull request by rip-nsk**
…to match the reader code.

### CODE REVIEW

**Comment by rip-nsk** [comment]
I'm not sure that it is correct fix, IMO, TimestampColumnReader should not do this:
        if (secsBuffer[i] < 0 && nanoBuffer[i] != 0) {
          secsBuffer[i] -= 1;
        }
but it seems necessary to match the java reader.

**Comment by omalley** [comment]
*Sigh* This is caused by the fact that Java defined Timestamp badly and then Hive copied those semantics into the ColumnVector.

We really need to fix these semantics in ORCv2.

I'll also comment that it would probably be nicer to our users if we weren't modifying the data in the columnVector they passed into us, but clearly we are already doing that.

File: c++/src/ColumnWriter.cc [hunk 1]
```diff
@@ -1199,6 +1199,10 @@ namespace orc {
         tsStats->increase(1);
         tsStats->update(millsUTC);
 
+        if (secs[i] < 0 && nanos[i] != 0) {
+          secs[i] += 1;
```
Comment (wgtmac) [inline c++/src/ColumnWriter.cc]: If secs[i] is -1 and we add 1 to it here. On the reader side, the if branch will simply fail to minus 1:
` if (secsBuffer[i] < 0 && nanoBuffer[i] != 0) { secsBuffer[i] -= 1; } `  This results in an 1 second shift of the original data. I canot find any good workaround but throw an exception if -1 has been found. @rip-nsk @omalley

Comment (rip-nsk) [reply c++/src/ColumnWriter.cc]: possible solutions 
- allow negative nanos in orc stream to cover the second before unix epoch.
- introduce new Timestamp type, w/o shift and deprecate old one

**Pull request closed without merging**